text/0000-opaque-data.md

```diff
@@ -0,0 +1,514 @@
+   the type).
+
+3. References and pointers to values of this type would be thin, having only a
+   single pointer width.
+
+4. A `Box<>` of this value would not free memory when dropped, like with ZSTs,
+   but would invoke the type's `drop(&mut self)` method if present.
+
+### Traits
+
+`Opaque` should implement the `Debug` trait, which should produce the string
+'Opaque'. It shouldn't implement any other traits from the standard library.
+
+## `size_of_val` and `align_of_val`
+
+Rust provides 2 intrinsics which are problematic to the implementation of
+`Opaque`, `size_of_val` and `align_of_val`. These two methods take an arbitrary
+unbounded reference, and return a `usize` representing the size of the value. We
+cannot implement these functions for `Opaque`, as it has a completely unknown
+layout, which in some cases may not be determinable even at runtime.
+
+There are multiple options for solving this problem, but this section will only
+cover one of them. The others are covered in the [Alternatives](#alternatives)
+section.
+
+## The `DynSized` trait
+
+A new built-in, unsafe, autoderived trait, `DynSized`, would be added. This type
+would be implemented on all existing types, except for `Opaque` and types which
+contain a `Opaque` object.
+
+Like `Sized`, the `DynSized` will be implied in all generic bounds, meaning that
+the generic declaration `<T>` would require the `DynSized` bound. Relaxing the
+`DynSized` bound would be done with `<T: ?DynSized>`.
+
+In addition, trait objects would be modified. Normal trait objects (`Trait`)
+would imply the `DynSized` bound, and contain the vtable entries required to
+implement the `size_of_val` and `align_of_val` intrinsics. Types which do not
+implement `DynSized` would not be able to be converted to a trait object of this
+kind. The syntax `Trait + ?DynSized` could also be written which would relax
+this `DynSized` requirement, and not include the vtable entries for
+`size_of_val` and `align_of_val`. `Sized` trait object should be able to
+implicitly coerce to `DynSized` trait objects by offseting their vtable pointer
+past the size and align vtable members.
+
+The `Sized` trait would be modified to inherit from the `DynSized` trait, meaning
+that `?DynSized` implies `?Sized`.
+
+Finally, many trait parmaeters would be updated in the standard library to relax
+their trait bounds from `?Sized` to `?DynSized` which should be a
+backwards-compatible change.
+
+`size_of_val` and `align_of_val` would require the `DynSized` trait.
+
+If we decide that we want to allow types which contain `Opaque` to specify a
+method for determining `size_of_val` and `align_of_val`, this could easily be
+done by allowing `DynSized` to be explicitly implemented:
+
+```rust
+import std::marker::DynSized;
+
+pub struct MyStruct(Opaque);
+
+unsafe impl DynSized for MyStruct {
+    fn size_of_val(&self) -> usize {
+        somehow_determine_size()
```

Comment:
I mostly mention the `impl DynSized` construct because it would be required in order to migrate `CStr` to using `Opaque`, rather than because it is something which I think is essential to the core of this RFC. I'd imagine that we would stabilize `impl DynSized` separately from the `Opaque` work, and mention as much in the RFC text.